docs(inputs): revamp events articles (#1912)
telerik/blazor-docs: components/datepicker/events.md

# Events

This article describes the events of the Telerik DatePicker for Blazor.

* [OnBlur](#onblur)
* [OnCalendarCellRender](#oncalendarcellrender)
* [OnChange](#onchange)
* [OnClose](#onclose)
* [OnOpen](#onopen)
* [ValueChanged](#valuechanged)

## OnBlur

The `OnBlur` event fires when the component loses focus.

>caption Handle the OnBlur event

````CSHTML
@* You do not have to use OnChange to react to loss of focus *@

@result

<TelerikDatePicker @bind-Value="@TheDate"
                   OnBlur="@OnBlurHandler">
</TelerikDatePicker>

@code {
    private string result = string.Empty;

    private DateTime? TheDate { get; set; } = DateTime.Today;

    private void OnBlurHandler()
    {
        result = string.Format("BLUR fired, current value is {0:dd/MMM/yyyy}", TheDate);
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## OnCalendarCellRender

The `OnCalendarCellRender` event fires when each calendar cell in each view is about to render. The event allows you to:
* Find out the current view.
* Find out the cell date.
* Set a custom CSS class for the `<td>` element.

> The event handler receives a `DatePickerCalendarCellRenderEventArgs` object as an argument, which contains the following properties:

Reviewer: ```suggestion
The event handler receives a [`DatePickerCalendarCellRenderEventArgs` object](/blazor-ui/api/telerik.blazor.components.datepickercalendarcellrendereventargs) as an argument, which contains the following properties:
```

>caption Handle the OnCalendarCellRender event.

````CSHTML
@* Customize the calendar cells using the OnCalendarCellRender event. *@

<TelerikDatePicker @bind-Value="@DatePickerValue"
                   OnCalendarCellRender="@OnCalendarCellRenderHandler"
                   Width="295px">
</TelerikDatePicker>

<style>
    .special {
        color: white;
        background-color: greenyellow;
        font-weight: bold;
    }
</style>

@code {
    private DateTime? DatePickerValue { get; set; } = DateTime.Today;

    private void OnCalendarCellRenderHandler(DatePickerCalendarCellRenderEventArgs args)
    {
        if (args.View == CalendarView.Month)
        {
            args.Class = args.Date.Day % 3 == 0 ? "special" : "";
        }
        else if (args.View == CalendarView.Decade)
        {
            args.Class = args.Date.Year == 2020 ? "special" : "";
        }
    }
}
````

## OnChange

The `OnChange` event represents a user action - confirmation of the current value. It fires when the user presses `Enter` in the input, or when the input loses focus.

> The DatePicker is a generic component, so you must provide either a `Value`, or a type to the `T` parameter of the component.

Reviewer: ```suggestion
The DatePicker is a generic component, so you must either provide a `Value`, or a type to the `T` parameter of the component.
```

# Events

This article describes the events of the Telerik DatePicker for Blazor.

* [OnBlur](#onblur)
* [OnCalendarCellRender](#oncalendarcellrender)
* [OnChange](#onchange)
* [OnClose](#onclose)
* [OnOpen](#onopen)
* [ValueChanged](#valuechanged)

## OnBlur

The `OnBlur` event fires when the component loses focus.

>caption Handle the OnBlur event

````CSHTML
@* You do not have to use OnChange to react to loss of focus *@

@result

<TelerikDatePicker @bind-Value="@DatePickerValue"
                   OnBlur="@OnBlurHandler">
</TelerikDatePicker>

@code {
    private string result = string.Empty;

    private DateTime? DatePickerValue { get; set; } = DateTime.Today;

    private void OnBlurHandler()
    {
        result = string.Format("BLUR fired, current value is {0:dd/MMM/yyyy}", DatePickerValue);
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## OnCalendarCellRender

The `OnCalendarCellRender` event fires when each calendar cell in each view is about to render. The event allows you to:

> * Find out the current view.

Reviewer 2: ```suggestion
* Identify the current view.
```

* Find out the cell date.
* Set a custom CSS class for the `<td>` element.

> The event handler receives a [`DatePickerCalendarCellRenderEventArgs` object](/blazor-ui/api/telerik.blazor.components.datepickercalendarcellrendereventargs) as an argument, which contains the following properties:

Reviewer 2: ```suggestion
As an argument, the event handler receives a [`DatePickerCalendarCellRenderEventArgs` object](/blazor-ui/api/telerik.blazor.components.datepickercalendarcellrendereventargs), which contains the following properties:
```

| `Date` | `DateTime` | The date of the calendar cell. |
| `View` | `CalendarView` enum <br /> (`Month`) | The currently visible view. You can use it to determine if the calendar is rendering the MonthView, YearView, and so on. |

>caption Handle the OnCalendarCellRender event.

````CSHTML
@* Customize the calendar cells using the OnCalendarCellRender event. *@

<TelerikDatePicker @bind-Value="@DatePickerValue"
                   OnCalendarCellRender="@OnCalendarCellRenderHandler"
                   Width="295px">
</TelerikDatePicker>

<style>
    .special {
        color: white;
        background-color: greenyellow;
        font-weight: bold;
    }
</style>

@code {
    private DateTime? DatePickerValue { get; set; } = DateTime.Today;

    private void OnCalendarCellRenderHandler(DatePickerCalendarCellRenderEventArgs args)
    {
        if (args.View == CalendarView.Month)
        {
            args.Class = args.Date.Day % 3 == 0 ? "special" : "";
        }
        else if (args.View == CalendarView.Decade)
        {
            args.Class = args.Date.Year == 2020 ? "special" : "";
        }
    }
}
````

## OnChange

> The `OnChange` event represents a user action - confirmation of the current value. It fires when the user presses `Enter` in the input, or when the input loses focus.

Reviewer 2: ```suggestion
The `OnChange` event represents a user action—confirmation of the current value. It fires when the user presses `Enter` in the input or when the input loses focus.
```

@* Cancel the OnClose event based on a condition *@

<TelerikDatePicker @bind-Value="@DatePickerValue"
                   OnClose="@OnDatePickerPopupClose">
</TelerikDatePicker>

@code {
    private DateTime? DatePickerValue { get; set; } = DateTime.Today;

    private void OnDatePickerPopupClose(DatePickerCloseEventArgs args)
    {
        //cancel the OnClose event based on a condition
        if (DatePickerValue > DateTime.Today)
        {
            args.IsCancelled = true;
        }
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## OnOpen

The `OnOpen` event fires before the DatePicker popup renders. 

> The event handler receives as an argument an `DatePickerOpenEventArgs` object that contains:

Reviewer: ```suggestion
The event handler receives a [`DatePickerOpenEventArgs` object](/blazor-ui/api/Telerik.Blazor.Components.DatePickerOpenEventArgs) as an argument, that contains the following properties:
```

> The event handler receives a [`DatePickerOpenEventArgs` object](/blazor-ui/api/Telerik.Blazor.Components.DatePickerOpenEventArgs) as an argument, that contains the following properties:

Reviewer 2: ```suggestion
As an argument, the event handler receives a [`DatePickerOpenEventArgs` object](/blazor-ui/api/Telerik.Blazor.Components.DatePickerOpenEventArgs), which contains the following properties:
```

@[template](/_contentTemplates/common/parameters-table-styles.md#table-layout)

| Property | Description |
| --- | --- |
| `IsCancelled` | Set the `IsCancelled` property to `true` to cancel the opening of the popup. |

````CSHTML
<TelerikDatePicker @bind-Value="@DatePickerValue"
                   OnOpen="@OnDatePickerPopupOpen">
</TelerikDatePicker>

@code {
    private DateTime? DatePickerValue { get; set; } = DateTime.Today;

    private void OnDatePickerPopupOpen(DatePickerOpenEventArgs args)
    {
        //set the IsCancelled to true to cancel the OnOpen event
        args.IsCancelled = false;
    }
}
````

@[template](/_contentTemplates/common/general-info.md#event-callback-can-be-async)

## ValueChanged

> The `ValueChanged` event fires upon every valid change in the input (for example, keystroke, after date selection in the Calendar popup).

Reviewer: ```suggestion
The `ValueChanged` event fires on every change in the component value, that occurs after typing or selection in the Calendar popup. If the `Value` parameter is nullable and the user types a value that is outside the `Min` and `Max` range, the handler will receive a `null` argument.
```

> The `ValueChanged` event fires on every change in the component value, that occurs after typing or selection in the Calendar popup. If the `Value` parameter is nullable and the user types a value that is outside the `Min` and `Max` range, the handler will receive a `null` argument.

Reviewer 2: ```suggestion
The `ValueChanged` event fires on every change in the component value that occurs after typing or selecting in the Calendar popup. If the `Value` parameter is nullable and the user types a value that is outside the `Min` and `Max` range, the handler will receive a `null` argument.
```
